### PR TITLE
Martin/ios 233 conversation list typing indicator

### PR DESCRIPTION
**Pull request by MartinJibber**

### CODE REVIEW

File: Jibber.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -12,7 +12,7 @@
 		7F227B8576E160C9469E08D1 /* Pods_Intents_Extension.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 2C50C41C9E4C24DF432E0D86 /* Pods_Intents_Extension.framework */; };
 		8007CB642703CE0600FEA511 /* AttachmentCollectionViewDataSource.swift in Sources */ = {isa = PBXBuildFile; fileRef = 8007CB632703CE0600FEA511 /* AttachmentCollectionViewDataSource.swift */; };
 		800FAAC126F0585A00D3C3A0 /* Conversation+Extensions.swift in Sources */ = {isa = PBXBuildFile; fileRef = 800FAAC026F0585A00D3C3A0 /* Conversation+Extensions.swift */; };
-		8015CA7726F3D907000F3C60 /* ConversationMessageCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 8015CA7626F3D907000F3C60 /* ConversationMessageCell.swift */; };
+		8015CA7726F3D907000F3C60 /* ConversationMessagesCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 8015CA7626F3D907000F3C60 /* ConversationMessagesCell.swift */; };
```
Comment (MartinJibber) [inline Jibber.xcodeproj/project.pbxproj:15]: Adding an "s" to the cell name for clarity.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 2]
```diff
@@ -10,15 +10,16 @@ import Foundation
...
+/// A cell to display a high-level view of a conversation's message. Displays a limited number of recent messages in a conversation.
+/// The user's messages and other messages are put in two stacks (along the z-axis),
+/// with the most recent message at the front (visually obscuring the others).
+class ConversationMessagesCell: UICollectionViewCell, ConversationMessageCellLayoutDelegate {
 
     // Interaction handling
-    var handleTappedMessage: ((Messageable) -> Void)?
-    var handleDeleteMessage: ((Messageable) -> Void)?
+    var handleTappedConversation: ((Messageable) -> Void)?
+    var handleDeleteConversation: ((Messageable) -> Void)?
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:20]: This cell now represents a whole conversation so you're really tapping on the whole conversation. Later on we can get more granular for opening up threads.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCellLayout.swift [hunk 3]
```diff
@@ -100,13 +100,15 @@ class ConversationMessageCellLayout: UICollectionViewFlowLayout {
...
-        = ChatClient.shared.messageController(cid: try! ChannelId(cid: recentMessage.conversationId),
-                                              messageId: recentMessage.id)
-        let mostRecentMessage = messageController.getMostRecent(fromCurrentUser: indexPath.section == 1)
+        guard let conversation = self.conversationDelegate.conversation else { return nil }
+
+        let cid = try! ChannelId(cid: conversation.conversationId)
+        let conversationController = ChatClient.shared.channelController(for: cid)
+        let mostRecentMessage
+        = conversationController.getMostRecentMessage(fromCurrentUser: indexPath.section == 1)
+
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCellLayout.swift:109]: Fixing issue where the date/time wasn't appearing.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController+Extensions.swift [hunk 4]
```diff
@@ -52,34 +52,34 @@ extension ConversationListViewController {
...
+        self.messageInputAccessoryView.textView.$inputText.mainSink { [unowned self] text in
+            guard let conversationController = self.conversationController else { return }
+
+            guard conversationController.areTypingEventsEnabled else { return }
+
+            if text.isEmpty {
+                conversationController.sendStopTypingEvent()
+            } else {
+                conversationController.sendKeystrokeEvent()
+            }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController+Extensions.swift:74]: Restoring the typing indicator events. The actual indicator isn't showing up yet.

File: iOS/_Stream Temp/ConversationController+Extensions.swift [hunk 5]
```diff
@@ -69,6 +69,20 @@ extension ChatChannelController {
...
+
+        // Find the most recent message that was sent by the user.
+        return allMessages.first { message in
+            if fromCurrentUser {
+                return message.isFromCurrentUser
+            } else {
+                return !message.isFromCurrentUser
+            }
+        }
+    }
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ConversationController+Extensions.swift:84]: Needed for show timestamps.

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**